File: leetcode-js/075-sort-colors.js
/**
 * @param {number[]} nums
 * @return {void} Do not return anything, modify nums in-place instead.
 */
var sortColors = function(nums) {
    let redIndex = 0;
    let blueIndex = nums.length - 1;
    let index = 0;

    while (index <= blueIndex) {
        if (nums[index] === 0) {
            const tmp = nums[redIndex];
            nums[redIndex] = nums[index];
            nums[index] = tmp;
            redIndex += 1;
            index += 1;
        } else if (nums[index] === 2) {
            const tmp = nums[blueIndex];
            nums[blueIndex] = nums[index];
            nums[index] = tmp;
            blueIndex -= 1;
        } else {
            index += 1;
        }
    }
};

File: leetcode-java/075-sort-colors.java
public class Solution {
    public void sortColors(int[] nums) {
        int r = 0;
        int b = nums.length - 1;
        int i = 0;
        while (i <= b) {
            if (nums[i] == 0) {
                swap(nums, r, i);
                r++;
                i++;
            } else if (nums[i] == 2) {
                swap(nums, b, i);
                b--;
            } else {
                i++;
            }
        }
    }

    void swap(int[] nums, int a, int b) {
        int t = nums[a];
        nums[a] = nums[b];
        nums[b] = t;
    }
}
